proposals/0570-string-interpolation.rst

```diff
@@ -0,0 +1,607 @@
+
+      textExample2 = T.pack $ T.unpack name1 <> " (age: " <> show age1 <> ") encountered " <> T.unpack name2 <> " (age: " <> show age2 <> ")"
+
+      textExample3 = T.unwords
+        [ name1
+        , "(age: " <> T.pack (show age1) <> ")"
+        , "encountered"
+        , name2
+        , "(age: " <> T.pack (show age2) <> ")"
+        ]
+
+* ``printf`` is partial and unsafe, which especially safety-conscious people might always stay away from anyway. Using a safer ``printf`` like ``formatting`` induces a third-party dependency, which is admittedly lightweight, but isn't as seamless of an integration as native string interpolation would be
+
+* Quasiquotes induces a dependency on Template Haskell, which a lot of people avoid out of principle. Most QuasiQuoters also add a dependency on ``haskell-src-exts`` to parse arbitrary Haskell expressions, which could technically be avoided by using something like ``ghc-meta`` (`repo <https://github.com/noughtmare/ghc-meta>`_, `GHC issue <https://gitlab.haskell.org/ghc/ghc/-/issues/20862>`_), but this isn't in wide use yet.
+
+If Haskell had native string interpolation, it would have the benefit and safety of the current third-party quasiquotes without the need for Template Haskell, and be able to take advantage of features like `multiline strings <https://github.com/ghc-proposals/ghc-proposals/pull/569>`_.
+::
+
+  error s"Expected: ${x + y}, got: ${result}"
+
+  let textExample = s"${name1} (age: ${age1}) encountered ${name2} (age: ${age2})"
+
+Proposed Change Specification
+-----------------------------
+
+This proposal introduces a new ``-XStringInterpolation`` extension that enables the following changes.
+
+Lexical Structure
+~~~~~~~~~~~~~~~~~
+
+Add ``interpolatedString`` and ``interpolatedStringMultiline`` to ``lexeme`` (not ``literal``, because they're not literals):
+
+.. code-block:: abnf
+
+  lexeme  → qvarid | qconid | qvarsym | qconsym
+          | literal | special | reservedop | reservedid
+          | interpolatedString
+          | interpolatedStringMultiline
+
+  interpolatedString → 's"' {graphic⟨'\' | '"' | '${'⟩ | space | escape | gap | '${' any⟨'}' | '"'⟩ '}'} '"'
+  interpolatedStringMultiline → 's"""' {{whitechar} interpolatedStringMultilineLine} '"""'
+  interpolatedStringMultilineLine → {graphic⟨'\' | '"""' | '${'⟩ | space | escape | gap | '${' ANY⟨'}' | '"'⟩ '}'}
+
+Also add ``$`` to ``charesc``:
+
+.. code-block:: abnf
+
+  charesc → a | b | f | n | r | t | v | \ | " | ' | & | $
+
+With ``$`` added to ``charesc``, interpolation can be avoided by escaping the dollar sign; e.g. ``s"\${foo}" == "${foo}"``.
+
+The following expressions are lex errors:
+
+* ``s"a ${"hello"} c"``
+
+  * This would lex the string ``s"a ${"``, then fail expecting a closing ``}``
+
+* ``s"a ${s"hello"} c"``
+
+  * This would lex the string ``s"a ${s"``, then fail expecting a closing ``}``
```

Comment:
Much better already. But reading up on https://peps.python.org/pep-0701/ (thanks for sticking our noses into it, @gkleen!), I think we might get away without the "re-lexing" pass, as follows.

My idea crucially builds on the fact that GHC's Lexer already maintains a stack of state numbers (simple integers), and the Lexer generator `alex` has special support for those as well.
Alas, it turns out that the string lexer is [hand-coded anyway (horrible indeed)](https://gitlab.haskell.org/ghc/ghc/-/blob/610840eb5bf6bd59417b82cc61e74aeaacbe5462/compiler/GHC/Parser/Lexer.x#L2184-2187), so we should not focus too much on the DFS declarations accepted by `alex`.

First off, keep `charesc` as is, but remove `interpolatedString`. (What I outline below always applies to `interpolatedMultilineString` as well.)

Second, define the token types

```
 isbegin → 's"'
 isend → '"'
 isinter → '$' '{'
 ismiddle → {graphic⟨'\' | '"' | isinter⟩ | space | escape | gap}
```

although `ismiddle` is nearly just `string⟨'$' '{'⟩`, but with additional quotes. perhaps  we should factor

```
  stringchar → graphic⟨'\' | '"'⟩ | space | escape | gap
  string → '"' {stringchar} '"'
  ismiddle → {stringchar⟨isinter⟩}
```

much better IMO.

The lexer is specified as follows, using two new lexer states `<istring>` and `<iexp>`:

1. After lexing `isbegin`, we push the new state `<istring>` onto the state stack that is active until the closing `isend` is lexed.
2. In state `<istring>`, only `ismiddle`, `isend` or `isinter` (the character sequence `${`) can be lexed.
3. When `isend` is lexed, we *pop* the lexer state and finish parsing of the string.
3. When `ismiddle` is lexed, we do not alter the lex state and remain in `<istring>`.
4. When `isinter` is lexed, we push a *new* lexer state `<iexp>`.
5. In state `<iexp>`, we basically proceed as in state `<0>`, which requires adding `iexp` to the respective lexer actions. There is one important difference, though: When we lex a token `}`, we pop the lexer state and go back into state `<istring>`.

With this, nested occurrences of `s"..${..}.."` should just lex correctly. The char literal '}' should lex correctly as well.
The only use case that is not supported is regular use of `}` in expressions, but that seems like a very minor drawback.

---

Here is the example from the PEP 701 adapted for our proposal:

```
s"some words ${x} more words ${'}'} final words"
```

this should result in the following token stream as per the lexer specified above:

```
isbegin - 's"'
ismiddle - 'some words '
isinter - '${'
qvarid - 'x'
ccurly - '}'
ismiddle - ' more words '
isinter - '${'
char - ''}''
ccurly - '}'
ismiddle - ' final words'
isend - '"'
```

---

The parser (section 2.2) will reconstruct a parse tree from this lexical syntax according to the following production rules:

```
isstring → isbegin { ismiddle | isinter exp '}' } isend
```

(I think the definition of `isend` is superfluous and we could just use the regular token `"`.)

---

I think this is much nicer, without any need for "re-lexing". 
The only question is: does the specification have an implementation (in `GHC.Parser.String.lexString`)?